### PR TITLE
Refactor pkg/oci

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This is a complete refactoring of the oci package. The whole idea is to implement all the OCI related functionalities we need in this package, avoiding other parts of the code having to interact with packages like "github.com/docker" or "oras.land/oras-go/v2".

The reasoning behind is:
- We can concentrate all OCI related code in the same place
- It's a lot easier to test if it's a separated package

This package now provides methods for pulling, pushing, tagging and building OCI images. Other methods that we previously exposed are now internal.

This commit also:
- Uses -ing in all error messages
- Small esthetic changes to cmd/*

NOTE: Ideally this should be multiple commits, but it's rather difficult to split a refactor like it into smaller pieces.

### CODE REVIEW

File: pkg/oci/oci.go [hunk 1]
```diff
@@ -281,22 +399,126 @@ func NewAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
...
-			return nil, fmt.Errorf("create auth client: %w", err)
+			return nil, fmt.Errorf("creating auth client: %w", err)
 		}
 		repo.Client = client
 	}
 
 	return repo, nil
 }
+
+func getImageListDescriptor(imageStore oras.ReadOnlyTarget, image string) (ocispec.Index, error) {
```
Comment (burak-ok) [inline pkg/oci/oci.go:422]:
```suggestion
func getImageListDescriptor(imageStore oras.ReadOnlyTarget, normalizedImage string) (ocispec.Index, error) {
```

From the codepath this function only gets normalized image names, so this is not an issue. The issue is when this function in the future gets called with a non normalized one.
So I think the least we should do to "catch" and future issues is to name the parameter that way (Or another way - you know the problems with naming things :D)

Now that I've thought about it a little more. Maybe we should introduce a new type. So the compiler always checks our usages of it. I just don't want headaches in the future xD. But I think this would be another PR

**Review by burak-ok — APPROVED** [review]
LGTM besides that (larger) comment that I have. But that should be another PR.

Tested it locally but probably didn't test all corner cases

File: pkg/oci/oci.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/oci/oci.go:422]: I see that parameter is passed to `imageStore.Resolve` and it's called there `reference`, so I think it's fine to call it reference. I'm not sure introducing a new type is a good idea, perhaps it adds a lot of boilerplate to cast from one to the other.

Comment (burak-ok) [reply pkg/oci/oci.go]: Yep, the boilerplate code is the disadvantage. And I also don't know how "annoying" or how much boilerplate would be needed.

File: pkg/oci/build.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+const (
+	ARCH_AMD64 = "amd64"
+	ARCH_ARM64 = "arm64"
+)
+
+type BuildGadgetImageOpts struct {
+	// List of eBPF objects to include on the image. The key is the architecture and the value
+	// is the path to the eBPF object.
+	EBPFObjectPaths map[string]string
+	// Path to the medatada file
```
Comment (flyth) [inline pkg/oci/build.go:43]:
```suggestion
	// Path to the metadata file
```

File: pkg/oci/build.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	"oras.land/oras-go/v2/errdef"
+)
+
+const (
+	ARCH_AMD64 = "amd64"
+	ARCH_ARM64 = "arm64"
+)
+
+type BuildGadgetImageOpts struct {
+	// List of eBPF objects to include on the image. The key is the architecture and the value
```
Comment (flyth) [inline pkg/oci/build.go:40]:
```suggestion
	// List of eBPF objects to include in the image. The key is the architecture and the value
```

File: pkg/oci/build.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	}
+	return nil
+}
+
+func createEbpfProgramDesc(target oras.Target, progFilePath string) (ocispec.Descriptor, error) {
+	progBytes, err := os.ReadFile(progFilePath)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("reading eBPF program file: %w", err)
+	}
+	progDesc := content.NewDescriptorFromBytes("application/vnd.gadget.ebpf.program.v1+binary", progBytes)
```
Comment (flyth) [inline pkg/oci/build.go:97]: Maybe use consts for `application/vnd.gadget.ebpf.program.v1+binary` and `application/vnd.gadget.config.v1+yaml` as well?

File: pkg/oci/build.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	// List of eBPF objects to include on the image. The key is the architecture and the value
+	// is the path to the eBPF object.
+	EBPFObjectPaths map[string]string
+	// Path to the medatada file
+	MetadataPath string
+}
+
+// BuildGadgetImage creates an OCI image with the objects provided in opts. The created image is
+// tagged if image is not empty.
+func BuildGadgetImage(opts *BuildGadgetImageOpts, image string) (*GadgetImageDesc, error) {
```
Comment (flyth) [inline pkg/oci/build.go:49]: Could you clarify a bit what exactly "The created image is tagged if image is not empty" means (maybe add an example?).

File: pkg/oci/build.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+func createEbpfProgramDesc(target oras.Target, progFilePath string) (ocispec.Descriptor, error) {
+	progBytes, err := os.ReadFile(progFilePath)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("reading eBPF program file: %w", err)
+	}
+	progDesc := content.NewDescriptorFromBytes("application/vnd.gadget.ebpf.program.v1+binary", progBytes)
+	progDesc.Annotations = map[string]string{
+		ocispec.AnnotationTitle:       "program.o",
+		ocispec.AnnotationAuthors:     "TODO: authors",
+		ocispec.AnnotationDescription: "TODO: description",
```
Comment (flyth) [inline pkg/oci/build.go:101]: Should these be exposed?

**Review by flyth — COMMENTED** [review]
Nice, I really like it!

Do you think it makes sense to promote the contexts to parameters where applicable now that you're refactoring anyway (and use `context.Background()` for now in the callers)?

File: pkg/oci/build.go (hunk 5 above)
Comment (mauriciovasquezbernal) [reply pkg/oci/build.go:49]: Added the comment in a different way. Not sure if the result is better.

File: pkg/oci/build.go (hunk 6 above)
Comment (mauriciovasquezbernal) [reply pkg/oci/build.go:101]: Yes, they should. However we're still understanding the different annotations to be used, so I prefer to skip that for now.

**Comment by mauriciovasquezbernal** [comment]
> Do you think it makes sense to promote the contexts to parameters where applicable now that you're refactoring anyway (and use `context.Background()` for now in the callers)?

Nice suggestion! I used context.TODO() in the caller to make it clear we still need to pass a valid context there later on.

File: pkg/oci/build.go (hunk 5 above)
Comment (flyth) [reply pkg/oci/build.go]: Yeah, that really helps, thanks!

**Review by flyth — APPROVED** [review]
LGTM, thanks!
